sdk/cosmosdb/azure_data_cosmos/src/clients/cosmos_client.rs

```diff
@@ -0,0 +1,81 @@
+use crate::authorization_policy::AuthorizationPolicy;
+use crate::{CosmosClientOptions, DatabaseClient};
+use azure_core::auth::TokenCredential;
+use azure_core::{Pipeline, Url};
+use std::sync::Arc;
+
+#[cfg(feature = "key-auth")]
+use azure_core::auth::Secret;
+
+/// Client for Azure Cosmos DB.
+#[derive(Debug, Clone)]
+pub struct CosmosClient {
+    endpoint: Url,
+    pub(crate) pipeline: Pipeline,
+
+    #[allow(dead_code)]
+    options: CosmosClientOptions,
+}
+
+pub trait CosmosClientMethods {
+    fn database(&self, name: impl Into<String>) -> DatabaseClient;
```

Comment:
Using an `impl` type here makes `CosmosClientMethods` **object-unsafe**. As long as we expect the trait to always be used as a generic constraint or in an `impl` type, that's OK, but it would prevent having a `Box<dyn CosmosClientMethods>` (or `Arc<dyn CosmosClientMethods>`, `&dyn CosmosClientMethods`, etc.).

The guidelines here are fairly clear though, [using an `impl` type is desirable](https://azure.github.io/azure-sdk/rust_introduction.html#rust-parameters) because it allows calling this like `client.database("foo")` rather than having to call `into()` yourself as a caller.

For the most part, we have this trait to support mocking. I think most applications will use `CosmosClient` directly, and those that care about mocking can use `impl CosmosClientMethods` or generic constraints (for example `Arc<T>` where `T: CosmosClientMethods` will work fine with this trait as is, it's just using `dyn` pointers that won't)

@heaths Is object-safety something that was considered when writing this guideline? As long as it's something we consider and consciously decide on, I'm alright with this trait being object-unsafe.

Reply:
No. I admit object-safety was not considered when writing these guidelines. The user experience was, and being able to test services is paramount. That said, if there are other considerations and we can be both object-safe and flexible, I'm open to it. As you noted, the main reason for preferring `Into<String>` (or `AsRef<str>` - whichever makes sense in context) over `&str` is flexibility but it's not like `&str` is that difficult. Callers can pass `"foo"` or `some_string.as_ref()` easily enough. I'm not thinking much about more abstract types that can be converted into a string e.g. integers.

Given we [can't declare `async fn`](https://doc.rust-lang.org/reference/items/traits.html#object-safety) for object-safe traits, I don't think it's worth considering `&str` in lieu of `Into<String>` or `AsRef<str>`, though.

/cc @RickWinter @JeffreyRichter

Reply:
I don't know all the ramifications around object-safety, so I think I'm not qualified to weigh in on this. I'm fine with approving this now and later having a discussion regarding the ramifications to make the best decision. As always, I'm hoping we get customer feedback which should help us either build confidence in what we've done or being flexible to change our decisions/designs.

Reply:
That makes sense. I generally agree that we don't need to design for object-safety here. The user experience is really much clearer if we can use `impl` types in the method names. I mostly just wanted to make sure that constraint was called out.